### PR TITLE
Incorperate delete() and download() into IUploader, update ResourceUp…

### PR DESCRIPTION
**Pull request by duttonw**
Fixes #
Incorporate delete() and download() into IUploader, update ResourceUpload, Upload with functionality from controller resource_download to allow better cloud integrations

### Proposed fixes:
Update IUploader method signagtures for: ResourceUpload, Upload to include delete() and download().

Update controller resource_download to utilise method

### Features:

- [ ] includes tests covering changes
- [x] includes updated documentation
- [ ] includes user-visible changes
- [x] includes API changes
- [x] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

File: ckan/logic/action/delete.py [hunk 1]
```diff
@@ -181,6 +182,13 @@ def resource_delete(context, data_dict):
         plugin.before_delete(context, data_dict,
                              pkg_dict.get('resources', []))
 
+    # Delete file if it was uploaded
+    if entity.get('url_type') == 'upload':
+        upload = uploader.get_resource_uploader(entity)
+        # Don't break if old plugin/interface is found
```
Comment (ThrawnCA) [inline ckan/logic/action/delete.py:188]: Perhaps log a warning if unsuccessful??

File: ckan/plugins/interfaces.py [hunk 2]
```diff
@@ -1717,6 +1731,20 @@ def get_resource_uploader(self):
         :param id: resource id
         :type id: string
 
+        ``delete(id)``
```
Comment (ThrawnCA) [inline ckan/plugins/interfaces.py:1734]: `filename` parameter should be shown.

File: ckan/plugins/interfaces.py [hunk 3]
```diff
@@ -1717,6 +1731,20 @@ def get_resource_uploader(self):
...
+
+        Perform a delete.
+
+        :param id: resource id, used to delete file
+        :param filename: can be None.
+
+        ``download(id)``
```
Comment (ThrawnCA) [inline ckan/plugins/interfaces.py:1741]: `filename` parameter should be shown.

File: ckan/plugins/interfaces.py [hunk 4]
```diff
@@ -1717,6 +1731,20 @@ def get_resource_uploader(self):
         :param id: resource id
         :type id: string
 
+        ``delete(id)``
+
+        Perform a delete.
+
+        :param id: resource id, used to delete file
+        :param filename: can be None.
```
Comment (ThrawnCA) [inline ckan/plugins/interfaces.py:1739]: A more comprehensive description would be nice, eg "The filename to use when storing a resource, may be None depending on the storage provider."

File: ckan/logic/action/delete.py (hunk 1 above)
Comment (duttonw) [reply ckan/logic/action/delete.py:188]:
```
# Delete file if it was uploaded
    if entity.get('url_type') == 'upload':
        upload = uploader.get_resource_uploader(entity)
        # Don't break if old plugin/interface is found
        if hasattr(upload, "delete"):
            upload.delete(id)
        else:
            logging.warning("%s does not have delete function, could not cleanup: %s", type(upload).__name__, id)
```

File: ckan/lib/uploader.py [hunk 5]
```diff
@@ -146,19 +148,21 @@ def update_data_dict(self, data_dict, url_field, file_field, clear_field):
         where the FieldStorage is kept (i.e the field where the file data
         actually is). clear_field is the name of a boolean field which
         requests the upload to be deleted.  This needs to be called before
-        it reaches any validators'''
+      x  it reaches any validators'''
```
[the remark this answers is not in the document]
Comment (duttonw) [reply ckan/lib/uploader.py]: removed

**Pull request merged**